src/util/sharing_map.h

```diff
@@ -299,100 +289,126 @@ class sharing_mapt
   void get_view(viewt &view) const;
 
   void get_delta_view(
-    const self_type &other,
+    const sharing_mapt &other,
     delta_viewt &delta_view,
-    const bool only_common=true) const;
+    const bool only_common = true) const;
 
 protected:
   // helpers
 
-  node_type *get_container_node(const key_type &k);
-  const node_type *get_container_node(const key_type &k) const;
+  innert *get_container_node(const key_type &k);
+  const innert *get_container_node(const key_type &k) const;
 
-  const node_type *get_leaf_node(const key_type &k) const;
+  const leaft *get_leaf_node(const key_type &k) const;
 
-  void gather_all(const node_type &n, delta_viewt &delta_view) const;
-};
+  void iterate(
+    const baset &n,
+    const unsigned depth,
+    std::function<void(const key_type &k, const mapped_type &m)> f) const;
 
-/// Get a view of the elements in the map
-/// A view is a list of pairs with the components being const references to the
-/// keys and values in the map.
-///
-/// Complexity:
-/// - Worst case: O(N * H * log(S))
-/// - Best case: O(N + H)
-///
-/// \param[out] view: Empty view
-SHARING_MAPT(void)::get_view(viewt &view) const
-{
-  assert(view.empty());
+  void gather_all(const baset &n, const unsigned depth, delta_viewt &delta_view)
+    const;
+
+  // dummy element returned when no element was found
+  static mapped_type dummy;
 
-  std::stack<const node_type *> stack;
+  static const std::string not_found_msg;
 
-  if(empty())
-    return;
+  // config
+  static const std::size_t bits;
+  static const std::size_t chunk;
+
+  // derived config
+  static const std::size_t mask;
+  static const std::size_t steps;
+
+  // key-value map
+  innert map;
 
-  stack.push(&map);
+  // number of elements in the map
+  size_type num = 0;
+};
+
+SHARING_MAPT(void)
+::iterate(
+  const baset &n,
+  unsigned depth,
+  std::function<void(const key_type &k, const mapped_type &m)> f) const
+{
+  typedef std::pair<unsigned, const baset *> stack_itemt;
+
+  std::stack<stack_itemt> stack;
+  stack.push({depth, &n});
 
   do
   {
-    const node_type *n=stack.top();
+    const stack_itemt &si = stack.top();
+
+    const unsigned depth = si.first;
+    const baset *bp = si.second;
+
     stack.pop();
 
-    if(n->is_container())
+    if(depth < steps) // internal
     {
-      for(const auto &child : n->get_container())
+      const innert *ip = static_cast<const innert *>(bp);
+      const to_mapt &m = ip->get_to_map();
+      SM_ASSERT(!m.empty());
+
+      for(const auto &item : m)
       {
-        view.push_back(view_itemt(child.get_key(), child.get_value()));
+        const innert *i = &item.second;
+        stack.push({depth + 1, i});
```

Comment:
You both push and pop on the stack within the do-while loop. Maybe comment on the property that ensures that the stack eventually becomes empty.